clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/core/ProjectClustersService.java

```diff
@@ -0,0 +1,415 @@
+ */
+package com.netflix.spinnaker.clouddriver.core;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.frigga.Names;
+import com.netflix.spinnaker.clouddriver.core.services.Front50Service;
+import com.netflix.spinnaker.clouddriver.model.Cluster;
+import com.netflix.spinnaker.clouddriver.model.ClusterProvider;
+import com.netflix.spinnaker.clouddriver.model.ServerGroup;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.inject.Provider;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class ProjectClustersService {
+
+  private final static Logger log = LoggerFactory.getLogger(ProjectClustersService.class);
+
+  private final Front50Service front50Service;
+  private final ObjectMapper objectMapper;
+  private final Provider<List<ClusterProvider>> clusterProviders;
+
+  public ProjectClustersService(Front50Service front50Service,
+                                ObjectMapper objectMapper,
+                                Provider<List<ClusterProvider>> clusterProviders) {
+    this.front50Service = front50Service;
+    this.objectMapper = objectMapper.copy().disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+    this.clusterProviders = clusterProviders;
+  }
+
+  public Map<String, List<ClusterModel>> getProjectClusters(List<String> projectNames) {
+    List<Map> projects = front50Service.searchForProjects(Collections.emptyMap(), 1000);
+
+    Map<String, List<ProjectClustersService.ClusterModel>> projectClusters = new ConcurrentHashMap<>();
+
+    for (Map projectMap : projects) {
+      String projectName = (String) projectMap.getOrDefault("name", "UNKNOWN");
+      if (!projectNames.contains(projectName)) {
+        continue;
+      }
+
+      Project project;
+      try {
+        project = objectMapper.convertValue(projectMap, Project.class);
+      } catch (IllegalArgumentException e) {
+        log.error("Could not marshal project '{}' to internal model", projectName, e);
+        continue;
+      }
+
+      if (project.config.clusters.isEmpty()) {
+        projectClusters.put(project.name, Collections.emptyList());
+        continue;
+      }
+
+      projectClusters.put(project.name, getProjectClusters(project));
+    }
+
+    return projectClusters;
+  }
+
+  public List<ClusterModel> getProjectClusters(String projectName) {
+    Map projectData = front50Service.getProject(projectName);
+
+    if (projectData == null) {
+      return null;
+    }
+
+    Project project;
+    try {
+      project = objectMapper.convertValue(projectData, Project.class);
+    } catch (IllegalArgumentException e) {
+      throw new MalformedProjectDataException("Could not marshal project to internal model: " + projectName, e);
+    }
+
+    return getProjectClusters(project);
```

Comment:
nitpick:

```
    try {
      return objectMapper.convertValue(projectData, Project.class);
    } catch (IllegalArgumentException e) {
      throw new MalformedProjectDataException("Could not marshal project to internal model: " + projectName, e);
    }
```

Reply:
Yet, I don't want a single project's malformed configuration to blow up the rest of the caching agent.

Reply:
Wat? Isn't my piece of code equivalent to your piece of code?

Reply:
Uhh... yes. 🤦‍♂️